core/src/main/java/org/infinispan/persistence/sifs/LogAppender.java

```diff
@@ -252,7 +252,12 @@ public void accept(WriteOperation writeOperation) {
             nextExpirationTime = -1;
             currentOffset = 0;
             logFile = null;
-            completeRequest(actualRequest);
+            // Completing in the completionProcessor ensures that we don't complete the clear until all completion
+            // notifications are done as well. This is important as those notifications update the index and we don't
+            // want to continue with clearing the index until all updates were sent to it
+            // This also ensure the complete log File update below is done before we clear as well - which would
+            // cause incorrect stats after a clear
+            completionProcessor.onNext(ignore -> completeRequest(actualRequest));
```

Comment:
Could the index update from writes be running concurrently on the index flowable? Would that be an issue?

Reply:
No, that isn't an issue as we don't clear the delete the log appender file until later when shutting down the FileProvider, so the contents will still be there for the index.

Note the order in https://github.com/infinispan/infinispan/blob/main/core/src/main/java/org/infinispan/persistence/sifs/NonBlockingSoftIndexFileStore.java#L486

So note the steps are
1. Pause and clear the log appender (any additional writes are enqueued until resumed) - also prevents additional index operations with this change
2. Compactor is cleared and paused (ignores any additional compaction requests) - also prevents additional index operations
3. Index sends a clear request down each index segment and completes after all are done.
4. Now with appender paused, the compactor and index can't have any more operations and we finally delete all the data directories from the FileProvider
5. Clear the temporary table since some index updates may be ignored as it was cleared (technically this can cause a blip where a value may be present, removed and then come back)
6. Finally resume the compactor and then log appender to resume operations

Note in all of these steps all of these things have ordering guaranteed by flowable to complete already submitted operations.

Regarding 5 above the entry will be present due to the temporary table then gone as it was removed in temporary table but not yet in the index until appender processes it again (if it was enqueued in the log appender). TBH, I don't know a good way to fix this, but also I don't how huge of an issue this is if someone is using clear with concurrent operations cannot expect the highest data consistency.